src/py21cmfast/src/Find_Haloes.c

```diff
@@ -0,0 +1,470 @@
+
+// Re-write of find_halos.c from the original 21cmFAST
+
+
+// Program ComputeHaloField takes in a k_space box of the linear overdensity field
+// and filters it on decreasing scales in order to find virialized halos.
+// Virialized halos are defined according to the linear critical overdensity.
+// ComputeHaloField outputs a cube with non-zero elements containing the Mass of
+// the virialized halos
+
+int check_halo(char * in_halo, struct UserParams *user_params, float R, int x, int y, int z, int check_type);
+int pixel_in_halo(struct UserParams *user_params, int x, int x_index, int y, int y_index, int z, int z_index, float Rsq_curr_index );
+
+int ComputeHaloField(float redshift, struct UserParams *user_params, struct CosmoParams *cosmo_params, struct AstroParams *astro_params, struct FlagOptions *flag_options, struct InitialConditions *boxes, struct HaloField *halos) {
```

Comment:
A lot of this function seems to be roughly the same as `ComputeInitialConditions` (without reading through and comparing explicitly). How much scope is there for refactoring out parts of the function to be used by both?

Reply:
It is actually quite different to `ComputeInitialConditions`, there are no redundant calculations between `ComputeHaloField` and `ComputeInitialConditions`. What might look similar is the generating wisdoms if they don't already exist.

That being said, I am toying with the idea of scrapping the halo classes altogether and attaching the halo lists directly to the `InitialConditions` and `PerturbField` classes. This would minimise the number of objects that need to be passed around to functions, requiring less modification of the existing framework.

I think I will make this change for this reason.

Reply:
OK cool, yeah that might be simpler overall.